addons/binding/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.dlinksmarthome.internal;
+
+import static org.openhab.binding.dlinksmarthome.DLinkSmartHomeBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jmdns.ServiceInfo;
+
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.io.transport.mdns.discovery.MDNSDiscoveryParticipant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DLinkSmartHomeDiscoveryParticipant} is responsible for discovering devices through UPnP.
+ *
+ * @author Mike Major - Initial contribution
+ *
+ */
+public class DLinkSmartHomeDiscoveryParticipant implements MDNSDiscoveryParticipant {
+
+    private static final String SERVICE_TYPE = "_dhnap._tcp.local.";
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return SUPPORTED_THING_TYPES_UIDS;
+    }
+
+    @Override
+    public String getServiceType() {
+        return SERVICE_TYPE;
+    }
+
+    @Override
+    public DiscoveryResult createResult(final ServiceInfo service) {
+        final ThingUID thingUID = getThingUID(service);
+
+        if (thingUID == null) {
+            return null;
+        }
+
+        final ThingTypeUID thingTypeUID = getThingType(service);
```

Comment:
The structure of this class was lifted from the chromecast binding. I had spotted the issue with getThingType being called more than once, but the getThingTypeUID method in ThingUID is marked as deprecated.

I've incorporated some of your suggestions, but the getThingUID method has to be implemented as this is part of the MDNSDiscoveryParticipant interface.